flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointFailureManager.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.checkpoint;
+
+import org.apache.flink.util.FlinkRuntimeException;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * The checkpoint failure manager which centralized manage checkpoint failure processing logic.
+ */
+public class CheckpointFailureManager {
+
+	private final static int UNLIMITED_TOLERABLE_FAILURE_NUMBER = Integer.MAX_VALUE;
+
+	private final int tolerableCpFailureNumber;
+	private final FailJobCallback failureCallback;
+	private final AtomicInteger continuousFailureCounter;
+
+	public CheckpointFailureManager(int tolerableCpFailureNumber, FailJobCallback failureCallback) {
+		checkArgument(tolerableCpFailureNumber >= 0
+				&& tolerableCpFailureNumber <= UNLIMITED_TOLERABLE_FAILURE_NUMBER,
+			"The tolerable checkpoint failure number is illegal, " +
+				"it must be greater than or equal to 0 and less than or equal to " + UNLIMITED_TOLERABLE_FAILURE_NUMBER + ".");
+		this.tolerableCpFailureNumber = tolerableCpFailureNumber;
+		this.continuousFailureCounter = new AtomicInteger(0);
+		this.failureCallback = checkNotNull(failureCallback);
+	}
+
+	/**
+	 * Handle checkpoint exception with a handler callback.
+	 *
+	 * @param exception the checkpoint exception.
+	 * @param checkpointId the failed checkpoint id used to count the continuous failure number based on
+	 *                     checkpoint id sequence. In trigger phase, we may not get the checkpoint id when the failure
+	 *                     happens before the checkpoint id generation. In this case, it will be specified a negative
+	 *                      latest generated checkpoint id as a special flag.
+	 */
+	public void handleCheckpointException(CheckpointException exception, long checkpointId) {
```

Comment:
I still see one important issue in this method that we should address: receiving multiple error reports for the same checkpoint id will continue to increase the count. For example, if the dfs goes away and suddenly all tasks report a problem for the same ongoing checkpoint, this would drive up the count, right? On checkpoint id should only be able to increase the count once. Special case is probably failing triggers that currently report negative checkpoint ids.

Reply:
Good catch! You are right, I will consider and fix it.

Reply:
I have added judge logic when counting, considering the failure reason of special case(negative checkpoint ids) are all ignored, it's OK we do not need to take care of it.